core/src/com/treode/async/misc/DirectDeque.scala

```diff
@@ -0,0 +1,95 @@
+package com.treode.async.misc
+
+import java.lang.System
+import scala.reflect.ClassTag
+
+/**
+ * DirectDeque is a queue (FIFO ordering) that supports
+ * constant time access to any element in the queue
+ */
+private class DirectDeque [M] (init_capacity: Int = 16) (implicit
+   mtag: ClassTag[M]
+) {
+  private var front = 0
+  private var back = front
+
+  private var array = new Array[M](init_capacity)
+
+  def size = {
+    if (front == back) {
+      0
+    } else if (front < back) {
+      back-front
+    } else {
+      (array.length-front)+back
+    }
+  }
+
+  def isEmpty() : Boolean = {
+    front == back
+  }
+
+ /**
+  * Adds element to the end of this DirectDeque
+  */
+  def enqueue (element: M) {
+    array(back) = element
+
+    if ((back+1)%array.length == front) {
+      array = resize()
+    } else {
+      back = (back+1)%array.length
+    }
+  }
+
+  /**
+   * Removes element from the front of this DirectDeque 
+   * in a first in first out (FIFO) order
+   * throws NoSuchElementException if there are no elements
+   */
+  def dequeue() : M = {
+    if (isEmpty()) {
+      throw new NoSuchElementException()
+    }
+
+    var retval = array(front)
+
+    front = (front+1)%array.length
+
+    retval
+  }
+
+  /**
+   * Returns but does not remove the nth element in
+   * this DirectDeque
+   * throws NoSuchElementException if n is < 0 or > number of elements
+   */
+  def get (n: Int) : M = {
+    if (n < 0 || n > size-1) {
+      throw new NoSuchElementException()
+    }
+
+    array((front+n)%array.length)
+  }
+
+  /**
+   * Returns a new array of double the size of the original array
+   * with the original elements copied into it
+   */
+  private def resize() : Array[M] = {
+    var prevlen = array.length
+    var newarray = new Array[M](2*prevlen)
+    
+    // copy elements from front till end of original array
+    System.arraycopy(array, front, newarray, 0, prevlen-front)
+    if (back < front) {
```

Comment:
First I'm copying from front to end of array (prevlen - front) and then I'm copying from 0 to back but since that will be one element short, I copy front number of elements because front = back + 1 if it is resizing.